### ISSUE TITLE
Support for setting option's control value to be a property of object and option's display value to be another property of the object

### ISSUE THREAD

**Issue by naveedahmed1** [angular/material2#14974]
I know that if we want the option's control value (what is saved in the form) to be different than the option's display value (what is displayed in the text field), we'll need to set the **displayWith** property on our autocomplete element. as explained here:

https://material.angular.io/components/autocomplete/overview#setting-separate-control-and-display-values

Suppose we have a user object like `{ id: 1, name: 'Abc' }`.

What if we want the option's control value (what is saved in the form) to be **user.id** and option's display value (what is displayed in the actual text field) to be **user.name**?

**Comment by naveedahmed1** [comment]
Yes, I know that's possible, but may we can have something like `displayWith` that allow specifying which value should be saved for form control, to avoid this step of assigning the entire object as the value and then only taking the specific property of object we want before firing off the request, may be something `saveValue`.

This would be really helpful for apps where we have dynamic forms, and where we just pass the model and the post url and rest is handles by the dynamic form component.

**Comment by crisbeto** [comment]
> Yes, I know that's possible, but may we can have something like displayWith that allow specifying which value should be saved for form control, to avoid this step of assigning the entire object as the value and then only taking the specific property of object we want before firing off the request, may be something saveValue.

The problem with this is that now the option will have two fields for the value, making it confusing for somebody that isn't familiar with the component.

**Comment by naveedahmed1** [comment]
What if we have a `saveWith` function similar to `displayWith`:

```
<mat-autocomplete #auto="matAutocomplete" [displayWith]="displayFn" [saveWith]="saveFn">
      <mat-option *ngFor="let option of filteredOptions | async" [value]="option">
        {{option.name}}
      </mat-option>
    </mat-autocomplete>
```

Or may be for `mat-option`, we could be able to specify display label:

```
<mat-autocomplete #auto="matAutocomplete" [displayWith]="displayFn">
      <mat-option *ngFor="let option of filteredOptions | async" [value]="option"  **[label]="option"**>
        {{option.name}}
      </mat-option>
    </mat-autocomplete>
```

By the way, I wonder why the options list behavior in `autocomplete` isn't similar to the options list in select list i.e. where when an option is selected, the option label is displayed for the user and the value of the option is saved in form control. If similar behavior could be implemented for `autocomplete`, this would be the best solution, I think.

**Comment by MRMokwa** [comment]
+1

This issue have already been asked before a couple of times, but didn't get much attention (the first reference I saw was about to get 2 years old).

For me this comment from @kdubious, summarize the problem: https://github.com/angular/material2/issues/8436#issuecomment-353203199

> I'm getting tripped up on implementing what I think is a fairly standard use case for a combobox (autocomplete).

For now, all we have is this jerry-rig, which solves the problem.
https://stackblitz.com/edit/angular-autocomplete-array-with-formcontrol-id-1?file=app%2Fautocomplete-display-example.ts

I would recommend avoiding use Autocomplete for this scenario, and use Selector componente instead. ... It's what I had to do unfortunatelly.

**Comment by muneebkq** [comment]
any update on this request
